### DISCUSSION TITLE
App crashes because of memory leakage while using .glb objects -> how to release the graphical memory allocated for the .glb based virtual objects

### DISCUSSION THREAD

**Discussion by Niloofar-didar** [sceneview/sceneform-android#216] [Q&A]
Hi, So I have been struggeling with the memory leak problem for months. I'm working on a project with android studio using AR-Core and Sceneform library (version 1.16). In this version I use .glb object format. The problem is that the graphical memory does not release the rendered objects after I choose to remove them and it causes the app to crash. I used the profiling tool of the android studio and I saw that the run-time memory usage keeps increasing till it reaches the maximum capacity (4GB) and the app crashes. The function to detach the virtual object has the bellow commands : { baseAnchor.getScene().onRemoveChild(baseAnchor.getParent()); baseAnchor.setRenderable(null); baseAnchor.setParent(null); SceneView.reclaimReleasedResources();}

the above commands just remove the object from the app visually but I beleive that the reference to the objects are still some where. The weird thing about this problem is that previously I used to have .SFB format object with the same detach function in sceneform 1.15 and the above commands worked there. The graphical memory is released for .SFB objects but in the new sceneform that supports gltf based objects the graphical memory is not released. Can some body help me with this problem? Thanks ahead

**Comment by Niloofar-didar** [comment]
I found SceneView.destroyAllResources(); as one command that originally gets into the renderer.java class of the sceneform library. In the code bellow you see that two functions are called:  ResourceManager.getInstance().destroyAllResources(); and EngineInstance.destroyEngine();
The first one tries to access the registry for rendering features which is referred as "resourceHolders" and release all the resources. The second one shuts down the engine and terminates the app. I altered the code to just call the first function without destroying the engine (commented this line) to see if the graphical mem is released but unfortunately nothing changed. The graphical mem is released only when we close the app. Is this problem related to the object type, i.e., because I'm using .glb file?

public static void destroyAllResources() {

    ResourceManager.getInstance().destroyAllResources();
   EngineInstance.destroyEngine();

  }



public void destroyAllResources() {
   for (ResourceHolder resourceHolder : resourceHolders) {
  resourceHolder.destroyAllResources();}

**Comment by fvito** [comment]
Hi,
From what I remember from the samples in this project, it did used to destroy the model resources when screen was closed, but there was an issue with the environment map, but I'm not sure if that is still the case. Can you try to reproduce the issue with one of the samples in the project? Also, are you maybe storing the renderables somewhere which makes then not be able to be reclaimed?

**Reply by Niloofar-didar** [reply]
Thanks for your response. Regarding the  destroyResource function, I manually used it in my code to see if the graphical resources would be released, and yes, this is called when you close the screen.
Regarding your second Qs, I don't store them somewhere that I can't access them. The problem should be with the sceneform. I think that it's not releasing the graphical memory for .glb objects

**Comment by ThomasGorisse** [comment]
You are using the 1.16 while we are currently on 1.20.1 !!!!

A lot of work has been made on the memory issues since the Google version

**Reply by Niloofar-didar** [reply]
Thanks Thomas for your response. I see the newer version in "https://github.com/SceneView/sceneform-android". I'll try it out and I hope this would solve the problem.

**Comment by Niloofar-didar** [comment]
Hello,
Oh, really! I didn’t know about the new version.
Could you please send me the link for the recent version? I hope this would solve the problem.
Thanks,
Niloofar

**Reply by RGregat** [reply]
Hi @niloofari I'm not aware about your current setup. But if you want to use the latest Version simple remove the old dependency to Sceneform and it this one 
```
dependencies {
     implementation("com.gorisse.thomas.sceneform:sceneform:1.20.2")
}
```

**Reply by Niloofar-didar** [reply]
Thank you so much. Will try that

**Comment by ThomasGorisse** [comment]
Here is the correct way to clean nodes:
[Remove or Hide a node](https://github.com/SceneView/sceneform-android#remove-or-hide-a-node)

**Reply by Niloofar-didar** [reply]
Thank you, I appreciate your help!

**Comment by ninjz** [comment]
@Niloofar-didar Did you make any progress on this? I'm experiencing issues with memory leak as well in my application. I'm calling all sorts of methods in my onDestroy handler like this and still no effect on my memory usage:
```java
    public void onDestroy() {
        characterModel.getRenderableInstance().getFilamentAsset().releaseSourceData();
        characterModel.getRenderableInstance().destroy();

        ResourceManager.getInstance().reclaimReleasedResources();
        ResourceManager.getInstance().destroyAllResources();

        SceneView.reclaimReleasedResources();
        sceneView.destroy();
        EngineInstance.getEngine().destroyScene(sceneView.getScene().getRenderer().getFilamentScene());
    }
```

**Reply by Niloofar-didar** [reply]
@ninjz No I couln't fix that and I believe that there is some bugs with the Sceneform Lib that handels the GLB object. I reverted my app back to the version with .OBJ and .SFB object file. With some experiment that I did to monitor the mem allocation and release in .SFB object models, I found that it can successfully release the memory after using the same function for removing the anchor node and  the reference to its parent. My siggestion for you is to try SFB models and use an older version of sceneform.

**Reply by ninjz** [reply]
@Niloofar-didar Thanks for your reply. Unfortunately I require the sceneform version which started supporting .GLB as the old sceneform does not support reverse animation playback.

**Reply by eqgis** [reply]
I fix that by modify this class.
In RenderableInstance.java.

 add this way to free memory.
```
/**
     * free gltf memo
     */
    public void destroyGltfAsset(){
        //desc- added by Ikkyu
        if (loader == null)return;
        if (filamentAsset != null){
            loader.destroyAsset(filamentAsset);
            filamentAsset = null;
        }
    }
```

Modify this method.
```
void createFilamentAssetModelInstance() {
        if (renderable.getRenderableData() instanceof RenderableInternalFilamentAssetData) {
            RenderableInternalFilamentAssetData renderableData =
                    (RenderableInternalFilamentAssetData) renderable.getRenderableData();
            Engine engine = EngineInstance.getEngine().getFilamentEngine();
            //updated by ikkyu
            loader =
                    new AssetLoader(
                            engine,
                            RenderableInternalFilamentAssetData.getMaterialProvider(),
                            EntityManager.get());
             //...
}
```

**Comment by ThomasGorisse** [comment]
Or just move to [SceneView](https://github.com/SceneView/sceneview-android) its quite easy if your project is not so big.

**Reply by ninjz** [reply]
I have tried to but ran into this issue in my project: https://github.com/SceneView/sceneview-android/issues/64 your help will be very much appreciated.

**Reply by ThomasGorisse** [reply]
I answered there : https://github.com/SceneView/sceneview-android/issues/64
